lib/base_intf.ml

```diff
@@ -48,11 +48,18 @@ module type S = sig
   val run : ('a, 'b) t -> 'a -> 'b
   (** [run r v] runs the reagents [r] with value [v]. *)
 
-
-  type catalyst
-
-  val catalyse : ('a, 'b) t -> 'a -> catalyst
-  val cancel_catalyst : catalyst -> unit
+  module Catalyst : sig
+    type catalyst
+    (** Type providing a handle for passively invoked reagent. Such a reagent 
+      remains active and available to react with other reagents until cancelled. *)
```

Comment:
Question: May a catalyst be automatically collected (and cancelled) as garbage (when `catalyst` handle is collected)?

Reply:
They kinda are at the moment. But their lifetime is tied to the structure on which they act rather than handle. 

We could tie catalysts' lifetimes to their handles (with a little hackery) but I think we would not want such a semantics in the general case. Some common uses I see for catalyst are specifying graphs of computations. For example doing a fan-in from multiple channels into one or joining different transformations together to create a pipeline. Right now if user drops the handle, they will keep an edge between two node active as long as source node is. They may also keep the handle to explicitly modify the graph again at some later time. That feels convenient, and there's no leak, because if source node (e.g. arrival queue) is collected, then all related catalysts will share its fate. 

Conversely, attaching the lifetimes to handles feels more error-prone. If I lose the handle by mistake, say it goes out of scope or I forget to copy it out of some record, at some point garbage collection cycle is going to dynamically cancel edge in my graph. Then some time later my arrival queue ends up overflowing.

Reply:
To add more color what happens inside: 

Reagents in non-blocking case are a thin-ish wrapper for summing sets of CASes. In the blocking case, things get a little more interesting. Reagents parametrise over scheduler to be able to suspend and resume threads at critical times rather than busy wait for others. 

This is reflected in the internals. Reagents have two modes: 
* Fast-path: just try the operation. If it produces a list of CASes to be execute, just commit those and we're done. 
* Blocking. If fast-path found that operation cannot finish (e.g. pop on empty stack), reagents' core generates an offer, which is published in relevant queue with some extra info and thread suspends on it.

Then, when another thread comes, it sees the offer and acts accordingly. In the case of `Ref` it's going to just wake-up waiters (hey, I've changed the value, you should retry). In the case of `Channel` it'll actually take suspended thread's transaction, merge it with their own, and try to commit everything at once. If the commit succeeds, it provides suspended thread with the result and resumes it. That cancels the offer.
 
Now, a catalyst is simply an offer that never expires. Whenever a new operation arrives, it has a partner to interact with, even though there's no actual thread waiting.

Reply:
Thanks!

In other words, the `catalyst` handle is specifically for explicitly canceling a catalyst rather than have it available for as long as the affected structure is alive.

Looking at this again, would it make sense to rename the type as `t`?  Then it would be `Catalyst.t` rather than `Catalyst.catalyst`.

Reply:
True that; renamed